src/server/cmd/pachctl/cmd/cmd.go

```diff
@@ -638,54 +639,47 @@ This resets the cluster to its initial state.`,
 	subcommands = append(subcommands, cmdutil.CreateAlias(portForward, "port-forward"))
 
 	var install bool
-	var path string
-	completion := &cobra.Command{
+	var installPathBash string
+	completionBash := &cobra.Command{
 		Short: "Print or install the bash completion code.",
-		Long:  "Print or install the bash completion code. This should be placed as the file `pachctl` in the bash completion directory (by default this is `/etc/bash_completion.d`. If bash-completion was installed via homebrew, this would be `$(brew --prefix)/etc/bash_completion.d`.)",
+		Long:  "Print or install the bash completion code.",
 		Run: cmdutil.RunFixedArgs(0, func(args []string) (retErr error) {
-			var dest io.Writer
-
 			if install {
-				f, err := os.Create(path)
-
-				if err != nil {
-					if os.IsPermission(err) {
-						fmt.Fprintf(os.Stderr, "Permission error installing completions, rerun this command with sudo.\n")
-					}
-					return err
+				if path.Base(installPathBash) != "pachctl" {
+					installPathBash = path.Join(installPathBash, "pachctl")
 				}
-
-				defer func() {
-					if err := f.Close(); err != nil && retErr == nil {
-						retErr = err
-					}
-
-					fmt.Printf("Bash completions installed in %s, you must restart bash to enable completions.\n", path)
-				}()
-
-				dest = f
-			} else {
-				dest = os.Stdout
 			}
-
-			// Remove 'hidden' flag from all commands so we can get bash completions for them as well
-			var unhide func(*cobra.Command)
-			unhide = func(cmd *cobra.Command) {
-				cmd.Hidden = false
-				for _, subcmd := range cmd.Commands() {
-					unhide(subcmd)
+			return createCompletions(rootCmd, install, installPathBash, rootCmd.GenBashCompletion)
+		}),
+	}
+	completionBash.Flags().BoolVar(&install, "install", false, "Install the completion.")
+	completionBash.Flags().StringVar(&installPathBash, "path", "/etc/bash_completion.d/pachctl", "Path to install the completion to.")
+	subcommands = append(subcommands, cmdutil.CreateAlias(completionBash, "completion bash"))
+
+	var installPathZsh string
+	completionZsh := &cobra.Command{
+		Short: "Print or install the zsh completion code.",
+		Long:  "Print or install the zsh completion code.",
+		Run: cmdutil.RunFixedArgs(0, func(args []string) (retErr error) {
+			if install {
+				if path.Base(installPathZsh) != "_pachctl" {
```

Comment:
Similarly here, I don't think we should be modifying the specified path like this?  Maybe if the destination exists and is a directory.